tests/test_staticfiles.py

```diff
@@ -279,3 +281,59 @@ def test_staticfiles_cache_invalidation_for_deleted_file_html_mode(tmpdir):
     )
     assert resp_deleted.status_code == 404
     assert resp_deleted.text == "<p>404 file</p>"
+
+
+def test_staticfiles_with_invalid_dir_permissions_returns_401(tmpdir):
+    path = os.path.join(tmpdir, "example.txt")
+    with open(path, "w") as file:
+        file.write("<file content>")
+
+    os.chmod(tmpdir, stat.S_IRWXO)
+
+    app = StaticFiles(directory=tmpdir)
+    client = TestClient(app)
+    response = client.get("/example.txt")
+    assert response.status_code == 401
+    assert response.text == "Permission denied"
+
+
+def test_staticfiles_with_missing_dir_returns_404(tmpdir):
+    path = os.path.join(tmpdir, "example.txt")
+    with open(path, "w") as file:
+        file.write("<file content>")
+
+    app = StaticFiles(directory=tmpdir)
+    client = TestClient(app)
+    response = client.get("/foo/example.txt")
+    assert response.status_code == 404
+    assert response.text == "Not Found"
+
+
+def test_staticfiles_access_file_as_dir_returns_404(tmpdir):
+    path = os.path.join(tmpdir, "example.txt")
+    with open(path, "w") as file:
+        file.write("<file content>")
+
+    app = StaticFiles(directory=tmpdir)
+    client = TestClient(app)
+    response = client.get("/example.txt/foo")
+    assert response.status_code == 404
+    assert response.text == "Not Found"
+
+
+def test_staticfiles_unhandled_os_error_returns_500(tmpdir):
+    path = os.path.join(tmpdir, "example.txt")
+    with open(path, "w") as file:
+        file.write("<file content>")
+
+    app = StaticFiles(directory=tmpdir)
+    client = TestClient(app)
+
+    with mock.patch(
+        "starlette.staticfiles.StaticFiles.lookup_path"
+    ) as mock_lookup_path:
```

Comment:
I haven't seen any usage of `unittest.mock` in the project, but I couldn't find any better way to create not-so-simple `OSError` exceptions.

Reply:
Is there any `OSError` that might cause this? Maybe a broken symlink or would that just 404?

Reply:
I guess that would be a FileNotFound. I'll give it another try but all other unhandled `OSError` exceptions happened on writing a file or internal sys calls.

Reply:
statting a file you can't access gives you PermissionError:

```
>>> os.stat("/boot/bar")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
PermissionError: [Errno 13] Permission denied: '/boot/bar'
```

Reply:
ah you already have a test for that

Reply:
Yeah, it looks like all other exceptions are network or sys call exceptions https://docs.python.org/3/library/exceptions.html#os-exceptions

Reply:
Yeah, I don't have any other ideas for testing this without a mock 🤔 

Would it be possible to use pytest's monkeypatch functionality for this? Wouldn't make much difference, but at least we'd be sticking to "one test framework". But if too much effort don't bother.

Reply:
@JayH5 I think that'd be better. I'll give it a try

Reply:
@JayH5 Changed that, please check again if that's what you meant.

Reply:
Thanks yes that is what I meant. Sorry for the slow response 😕 

I've been thinking about this again and I'm not sure we actually want to return 500 from within the static files implementation. Should we not just allow/raise the exception since it is so unexpected? IMO, the `ServerErrorMiddleware` (and similar) should really be the only thing in Starlette that returns a 500 response.

Reply:
@JayH5
That makes sense. What about the other exceptions?
Like 404 and 401, can't `ExceptionMiddleware` handle those?

Reply:
You do make a good point 🤔 I'm not sure why we don't raise an `HTTPException` in those cases. I'd be open to that change (if it's as straightforward as it seems) either in this PR or another.

Reply:
@JayH5  I played around this a bit and changed that.
I think we can raise `HTTPException` instead of returning `PlaintTextResponse`s but there's a difference. If we raise `HTTPException` then we need to can't use `StaticFiles` alone as we need `Starlette` middlewares to handle the exception which `StaticFiles` doesn't have.